incubator/hnc/pkg/controllers/object_controller.go

```diff
@@ -104,68 +103,93 @@ func (r *ObjectReconciler) SyncNamespace(ctx context.Context, log logr.Logger, n
-// it's correct and ready to be propagated. The incorrect object will be deleted
-// if it's obsolete; or marked as modified if it's changed.
-func (r *ObjectReconciler) ensureCorrectAncestry(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, error) {
-	// Find what we expect the source to be.
-	missing, src, err := r.getSourceInst(ctx, log, inst)
-	if err != nil {
-		return false, err
+// syncWithForest syncs the object instance with the in-memory forest. This is the only place we
+// should lock the forest, so this fn needs to return everything relevant for the rest of the
+// reconciler. It returns true if the object instance is good and shouldn't be deleted. It will
+// also return a list of destinations for propagation.
+func (r *ObjectReconciler) syncWithForest(ctx context.Context, log logr.Logger, inst, srcInst *unstructured.Unstructured) ([]string, bool) {
+	r.Forest.Lock()
+	defer r.Forest.Unlock()
+
+	if !r.hasCorrectAncestry(ctx, log, inst, srcInst) {
+		return nil, false
 	}
 
-	// The source instance no longer exists in the source namespace.
-	if missing {
-		log.Info("Will delete because the source object no longer exists in ancestor namespace(s)", "inheritedFrom", inst.GetLabels()[labelInheritedFrom])
-		return false, r.Delete(ctx, inst)
+	if !r.isUnmodified(ctx, log, inst, srcInst) {
+		// TODO add object condition here to replace the setAnnotation
+		log.Info("Marking as modified from the source", "inheritedFrom", srcInst.GetNamespace())
+		// Mark as modified if the copied fields don't match.
+		setAnnotation(inst, annotationModified, "true")
+		// TODO it's wrong to call apiserver in forest lock, will be replaced soon.
+		err := r.Update(ctx, inst)
+		if err != nil {
+			log.Error(err, "Couldn't add modified annotation")
+		}
+		return nil, true
 	}
 
-	// If there's no source, *this* is the source
+	return r.Forest.Get(inst.GetNamespace()).ChildNames(), true
+}
+
+// hasCorrectAncestry checks to see if the given object has correct ancestry. It returns false
+// if the source namespace is no longer an ancestor and the object should be deleted.
+func (r *ObjectReconciler) hasCorrectAncestry(ctx context.Context, log logr.Logger, inst, src *unstructured.Unstructured) bool {
 	if src == nil {
-		return true, nil
+		log.Info("Will delete because the source instance no longer exists", "inheritedFrom", getSourceNS(inst))
+		return false
+	}
+
+	if inst == src {
+		return true
+	}
+
+	// The source exists. Is it still in an ancestor namespace?
+	curNS := r.Forest.Get(inst.GetNamespace())
+	srcNS := r.Forest.Get(src.GetNamespace())
+	if !curNS.IsAncestor(srcNS) {
+		log.Info("Will delete because the source namespace is no longer an ancestor", "inheritedFrom", srcNS)
+		return false
+	}
+
+	return true
+}
+
+// isUnmodified checks to see if this object is the same as the source. It returns true if it's
+// correct and ready to be propagated.
```

Comment:
I don't think you need the second sentence anymore. The _caller_ decides that this is ready to be propagated, not this function - this function literally just decides whether it's modified or not.